### PR TITLE
Support alternative stack for signal handling

### PR DESCRIPTION
**Pull request by jxyang**
Signed-off-by: Xuejun Yang <xuejya@microsoft.com>

### CODE REVIEW

File: solutions/coreclr/test-runner.sh [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 USAGE="$0 path-to-myst exec|exec-sgx|exec-linux ext2|cpio <test_list_file>"
 NPROCS=$(grep -c ^processor /proc/cpuinfo)
-#NPROCS=1
+NPROCS=2
```
Comment (vtikoo) [inline solutions/coreclr/test-runner.sh:5]: Any reason to reduce the parallelism?

File: kernel/signal.c [hunk 2]
```diff
@@ -437,3 +491,47 @@ long myst_handle_host_signal(siginfo_t* siginfo, mcontext_t* mcontext)
...
+
+    if (ss != NULL)
+    {
+        // We don't support swapcontext in libc. No point
+        // to support SS_AUTODISARM which was introduced to enable it.
+        // The only flag allowed here is SS_DISABLE.
+        if (ss->ss_flags & ~(SS_DISABLE) != 0)
+            ERAISE(-EINVAL);
+
+        if (ss->ss_size < MINSIGSTKSZ)
```
Comment (vtikoo) [inline kernel/signal.c:517]: Should this be checked after handling SS_DISABLE? Acc to man page - 
>  In this case, the kernel ignores any other flags in ss.ss_flags and the remaining fields in ss.

File: kernel/signal.c [hunk 3]
```diff
@@ -238,27 +271,48 @@ static long _handle_one_signal(
...
         {
-            ((sigaction_handler_t)(action->handler))(signum);
+            arg.signum_handler = (sigaction_handler_t)(action->handler);
+            arg.signum = signum;
         }
 
+        if (use_alt_stack)
+        {
+            uint64_t stacktop = (uint64_t)altstack->ss_sp + altstack->ss_size;
+            altstack->ss_flags |= SS_ONSTACK;
```
Comment (vtikoo) [inline kernel/signal.c:303]: Can you add a comment here on why adding this flag is needed during handling signal on alternate stack.

File: solutions/coreclr/test-runner.sh (hunk 1 above)
Comment (jxyang) [reply solutions/coreclr/test-runner.sh:5]: Seems reduce the parallelism doesn't improve the speed. Will revert back.

File: kernel/signal.c (hunk 2 above)
Comment (jxyang) [reply kernel/signal.c:517]: Thanks for catching that. Will move.

File: kernel/signal.c (hunk 3 above)
Comment (jxyang) [reply kernel/signal.c:303]: Added

Comment (vtikoo) [reply kernel/signal.c:307]: Thanks. This is also checked in `myst_signal_altstack` if user tries to change the alternate stack while being on it.

**Review by vtikoo — APPROVED** [review]
LGTM. Please also update the relevant ltp tests and syscall limitation doc.

File: kernel/signal.c [hunk 4]
```diff
@@ -437,3 +495,46 @@ long myst_handle_host_signal(siginfo_t* siginfo, mcontext_t* mcontext)
...
+        {
+            if (ss->ss_size < MINSIGSTKSZ)
+                ERAISE(-ENOMEM);
+            me->signal.altstack = *ss;
+        }
+    }
+
+done:
+    return ret;
+}
```
Comment (mikbras) [inline kernel/signal.c:540]: Please run code-format on this file.

File: kernel/signal.c [hunk 5]
```diff
@@ -238,27 +271,52 @@ static long _handle_one_signal(
...
+        if (use_alt_stack)
+        {
+            uint64_t stacktop = (uint64_t)altstack->ss_sp + altstack->ss_size;
+            // Remember this thread is already on the alt stack. We check
+            // this flag when the next signal comes in, if true we will
+            // continue on the current (alt) stack instead of starting
+            // from the top of the alt stack.
+            altstack->ss_flags |= SS_ONSTACK;
+
+            myst_call_on_stack((void*)stacktop, _handler_wrapper, &arg);
```
Comment (mikbras) [inline kernel/signal.c:309]: Does the handler necessarily return? Can it do a longjmp?

Comment (mikbras) [inline kernel/signal.c:309]: The return value of ``_handler_wrapper()`` is not checked but I see that it returns 0, so that is okay.

**Review by mikbras — APPROVED** [review]

**Pull request merged**